src/test/chain_tests.cpp

```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    BOOST_CHECK_EQUAL(GetSkipHeight(2), 0);
+    BOOST_CHECK_EQUAL(GetSkipHeight(3), 1);
+    BOOST_CHECK_EQUAL(GetSkipHeight(4), 0);
+    BOOST_CHECK_EQUAL(GetSkipHeight(5), 1);
+    BOOST_CHECK_EQUAL(GetSkipHeight(6), 4);
+    BOOST_CHECK_EQUAL(GetSkipHeight(7), 1);
+    BOOST_CHECK_EQUAL(GetSkipHeight(8), 0);
+    BOOST_CHECK_EQUAL(GetSkipHeight(9), 1);
+    BOOST_CHECK_EQUAL(GetSkipHeight(10), 8);
+}
+
+BOOST_AUTO_TEST_CASE(skip_height_analysis)
+{
+    // This test case indirectly tests the `GetSkipHeight` function,
+    // and asserts that its properties are suitable for performant
+    // ancestor search.
+    // It does so by imitating the `GetAncestor` function to
+    // look for an ancestor, and ensuring that an arbitrary
+    // ancestor is found in a number of steps that is much lower
+    // than simple linear traversal.
+
+    // Build a chain
+    const auto chain_size{200'000};
+    std::vector<std::unique_ptr<CBlockIndex>> block_index;
+    block_index.reserve(chain_size);
+    // Create genesis block
+    block_index.push_back(std::make_unique<CBlockIndex>());
+    for(auto i{1}; i < chain_size; ++i) {
+        auto new_index{std::make_unique<CBlockIndex>()};
+        new_index->pprev = block_index.back().get();
+        BOOST_CHECK(new_index->pprev);
+        new_index->nHeight = new_index->pprev->nHeight + 1;
+        new_index->BuildSkip();
+        block_index.push_back(std::move(new_index));
+    }
+
+    const auto runs{1000};
+    const auto min_distance{500};
+    FastRandomContext ctx(/*fDeterministic =*/true);
+    for (auto i{0}; i < runs; ++i) {
+        // pick a height in the second half of the chain
+        auto start_height = chain_size / 2 + ctx.randrange(chain_size / 2);
+        BOOST_REQUIRE(start_height > 0 && start_height < chain_size);
+        auto& start_p = block_index[start_height];
+        // pick a target height, earlier by at least min_distance
+        auto delta{min_distance + ctx.randrange(start_height - min_distance)};
+        auto target_height{start_height - delta};
+        BOOST_REQUIRE(target_height > 0 && target_height < chain_size && target_height < start_height);
+
+        // Perform traversal from start to target, skipping, as implemented in `GetAncestor`
+        const CBlockIndex* pindexWalk = start_p.get();
+        auto heightWalk = start_height;
+        auto iteration_count{0};
+        while (heightWalk > target_height) {
+            int heightSkip = GetSkipHeight(heightWalk);
+            int heightSkipPrev = GetSkipHeight(heightWalk - 1);
+            if (pindexWalk->pskip != nullptr &&
+                (heightSkip == target_height ||
+                (heightSkip > target_height && !(heightSkipPrev < heightSkip - 2 &&
+                                        heightSkipPrev >= target_height)))) {
+                // Only follow pskip if pprev->pskip isn't better than pskip->pprev.
+                pindexWalk = pindexWalk->pskip;
+                heightWalk = heightSkip;
```

Comment:
we should assert that this will always converge towards genesis (i.e. it's decreasing):
```C++
                BOOST_REQUIRE(skip->nHeight < height_walk);
```

Reply:
Added now.